pkgs/development/libraries/gaia/default.nix

```diff
@@ -0,0 +1,81 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, libyaml
+, swig
+, pkgconfig
+, wafHook
+, makeWrapper
+, qt4
+, python
+, pythonSupport ? true
+# Default to false since it brakes the build
+, stlfacadeSupport ? false
+, assertsSupport ? true
+, cyclopsSupport ? true
+}:
+
+assert pythonSupport -> python != null;
+
+stdenv.mkDerivation rec {
+  pname = "gaia";
+  version = "2.4.5";
+
+  src = fetchFromGitHub {
+    owner = "MTG";
+    repo = "gaia";
+    rev = "v${version}";
+    sha256 = "12jxb354s2dblr2ghnl3w05m23jgzvrrgywfj8jaa32j3gw48fv2";
+  };
+
+  # Fix installation error when waf tries to put files in /etc/
+  prePatch = ''
+  '' + lib.optionalString cyclopsSupport ''
+    substituteInPlace src/wscript \
+      --replace "/etc/cyclops" "$out/etc/cyclops" \
+      --replace "/etc/init.d" "$out/etc/init.d"
+  '';
+
+  # This is not exactly specified in upstream's README but it's needed by the
+  # resultings $out/bin/gaiafusion script
+  pythonEnv = (if pythonSupport then
+    python.withPackages(ps: with ps; [
+      pyyaml
+    ])
+  else null);
+
+  nativeBuildInputs = [
+    wafHook
+    pkgconfig
+    swig
+    makeWrapper
+  ];
+  buildInputs = [
+    libyaml
+    qt4
+  ]
+    ++ lib.optionals (pythonSupport) [
+      pythonEnv
+    ]
+  ;
+  wafConfigureFlags = [
+  ]
+    ++ lib.optionals (pythonSupport) [ "--with-python-bindings" ]
+    ++ lib.optionals (stlfacadeSupport) [ "--with-stlfacade" ]
+    ++ lib.optionals (assertsSupport) [ "--with-asserts" ]
+    ++ lib.optionals (cyclopsSupport) [ "--with-cyclops" ]
+  ;
+  # only gaiafusion is a python executable that needs patchShebangs
+  postInstall = (if pythonSupport then ''
+    # We can't use patchShebangs because it will use bare bones $python/bin/python
+    # and we need a python environment with pyyaml
+    wrapProgram $out/bin/gaiafusion --prefix PYTHONPATH : $out/${python.sitePackages}:${pythonEnv}/${python.sitePackages}
```

Comment:
Why not use `wrapPython` instead? Using an extra python environment seems overkill if you just need `pyyaml`.

Reply:
Sounds like a good idea, I didn't know this existed. I wouldn't mind giving up upon the `pythonEnv` and use the dependencies straight in the inputs. However, whether I use a `pythonEnv` or not, the harder part is that the `$out/bin/gaiafusion` script requires the `$PYTHONPATH` to include gaia's python modules - those in `$out/${python.sitePackages}`.

I tried the following:

```patch
diff --git i/pkgs/development/libraries/gaia/default.nix w/pkgs/development/libraries/gaia/default.nix
index dcf44ec6ca4..5a060f5001b 100644
--- i/pkgs/development/libraries/gaia/default.nix
+++ w/pkgs/development/libraries/gaia/default.nix
@@ -36,26 +36,25 @@ stdenv.mkDerivation rec {
       --replace "/etc/init.d" "$out/etc/init.d"
   '';
 
-  # This is not exactly specified in upstream's README but it's needed by the
-  # resultings $out/bin/gaiafusion script
-  pythonEnv = (if pythonSupport then
-    python.withPackages(ps: with ps; [
-      pyyaml
-    ])
-  else null);
-
   nativeBuildInputs = [
     wafHook
     pkgconfig
     swig
     makeWrapper
-  ];
+  ]
+    # The gaiafusion binary inside $out/bin needs a shebangs patch, and
+    # wrapping with the appropriate $PYTHONPATH
+    ++ lib.optionals (pythonSupport) [ pythonPackages.wrapPython ]
+  ;
   buildInputs = [
     libyaml
     qt4
   ]
     ++ lib.optionals (pythonSupport) [
-      pythonEnv
+      pythonPackages.python
+      # This is not exactly specified in upstream's README but it's needed by the
+      # resulting $out/bin/gaiafusion script
+      pythonPackages.pyyaml
     ]
   ;
   wafConfigureFlags = [
@@ -65,12 +64,6 @@ stdenv.mkDerivation rec {
     ++ lib.optionals (assertsSupport) [ "--with-asserts" ]
     ++ lib.optionals (cyclopsSupport) [ "--with-cyclops" ]
   ;
-  # only gaiafusion is a python executable that needs patchShebangs
-  postInstall = lib.optionalString ''
-    # We can't use patchShebangs because it will use bare bones $python/bin/python
-    # and we need a python environment with pyyaml
-    wrapProgram $out/bin/gaiafusion --prefix PYTHONPATH : $out/${python.sitePackages}:${pythonEnv}/${python.sitePackages}
-  '';
 
   meta = with lib; {
     homepage = "https://github.com/MTG/gaia";
```

But `$out/bin/gaiafusion` comes out as if I had used a bare `patchShebangs` - it doesn't find gaia's modules.

I tried also using a `postInstall` including `wrapPythonPrograms` but then `$out/bin/gaiafusion` couldn't find `yaml`.

I tried also using `wrapPythonProgramsIn $out/bin/ $PYTHONPATH:$out/${pythonPackages.python.sitePackages}` but then `$out/bin/gaiafusion` couldn't find gaia's modules as well.

I noticed that the shell function `wrapPythonProgramsIn` doesn't send `$PYTHONPATH` to `wrapProgram`. See:

https://github.com/NixOS/nixpkgs/blob/a3f24daa7bc52767e15b5476d6c6b4165db5ee39/pkgs/development/interpreters/python/wrap.sh#L71-L77

I'd expect that shell array to include:

```
--prefix PYTHONPATH : "$program_PYTHONPATH"
```

I tried adding these strings to the array but that turned out to trigger a massive rebuild of almost every package so I gave up. I did try to use `makeWrapperArgs` in the derivation:

```nix
  makeWrapperArgs = [
    "--prefix" "PYTHONPATH" ":" "$program_PYTHONPATH"
  ];
```

Hoping `wrapPythonProgramsIn` will consider it when writing `wrapProgram`. But then `$out/bin/gaiafusion` didn't find gaia's modules as well :disappointed:.

It's kind of disappointing none of these shell functions work and that I have to use the good old `wrapProgram` my self. It's not that bad TBH, but it's not as elegant as well.

Reply:
Ah, I should have been more clear on this, here is a diff of it working:

```diff
diff --git a/pkgs/development/libraries/gaia/default.nix b/pkgs/development/libraries/gaia/default.nix
index 7565ee7a06e..9367792e06c 100644
--- a/pkgs/development/libraries/gaia/default.nix
+++ b/pkgs/development/libraries/gaia/default.nix
@@ -36,41 +36,23 @@ stdenv.mkDerivation rec {
       --replace "/etc/init.d" "$out/etc/init.d"
   '';
 
-  # This is not exactly specified in upstream's README but it's needed by the
-  # resulting $out/bin/gaiafusion script
-  pythonEnv = (if pythonSupport then
-    pythonPackages.python.withPackages(ps: with ps; [
-      pyyaml
-    ])
-  else null);
-
   nativeBuildInputs = [
     wafHook
     pkgconfig
     swig
     makeWrapper
-  ];
-  buildInputs = [
-    libyaml
-    qt4
-  ]
-    ++ lib.optionals (pythonSupport) [
-      pythonEnv
-    ]
-  ;
-  wafConfigureFlags = [
-  ]
-    ++ lib.optionals (pythonSupport) [ "--with-python-bindings" ]
-    ++ lib.optionals (stlfacadeSupport) [ "--with-stlfacade" ]
-    ++ lib.optionals (assertsSupport) [ "--with-asserts" ]
-    ++ lib.optionals (cyclopsSupport) [ "--with-cyclops" ]
-  ;
-  # only gaiafusion is a python executable that needs patchShebangs
-  postInstall = lib.optionalString (pythonSupport) ''
-    # We can't use patchShebangs because it will use bare bones $python/bin/python
-    # and we need a python environment with pyyaml
-    wrapProgram $out/bin/gaiafusion --prefix PYTHONPATH : $out/${pythonPackages.python.sitePackages}:${pythonEnv}/${pythonPackages.python.sitePackages}
-  '';
+  ] ++ lib.optional pythonSupport pythonPackages.wrapPython;
+
+  buildInputs = [ libyaml qt4 ];
+
+  propagatedBuildInputs = lib.optional pythonSupport pythonPackages.pyyaml;
+
+  wafConfigureFlags = lib.optional pythonSupport "--with-python-bindings"
+                   ++ lib.optional stlfacadeSupport "--with-stlfacade"
+                   ++ lib.optional assertsSupport "--with-asserts"
+                   ++ lib.optional cyclopsSupport "--with-cyclops";
+
+  postFixup = "wrapPythonPrograms";
 
   meta = with lib; {
     homepage = "https://github.com/MTG/gaia";
```

The point here is that wrapping Python programs is only done automatically if you actually use the `buildPython*` functions. What's different in the above is the additional `postFixup` hook that's calling `wrapPythonPrograms`, which then will add all Python inputs in `propagatedBuildInputs` to the corresponding executables.

Reply:
I see. Thanks for the patch! I've applied a version of mine to the branch and pushed the changes. The only differences are that I used `lib.optionals` and other styles of writing we are do not agree on :smile:.